crates/bevy_ecs/src/observer/builder.rs

```diff
@@ -0,0 +1,233 @@
+use std::any::TypeId;
+
+use super::*;
+
+/// Builder struct for [`Observer`].
+pub struct ObserverBuilder<'w, E = ()> {
+    commands: Commands<'w, 'w>,
+    descriptor: ObserverDescriptor,
+    _marker: PhantomData<E>,
+}
+
+impl<'w, E: 'static> ObserverBuilder<'w, E> {
+    /// Constructs a new [`ObserverBuilder`].
+    pub fn new(commands: Commands<'w, 'w>) -> Self {
+        let mut descriptor = ObserverDescriptor::default();
+        let event = commands
+            .components()
+            .get_id(TypeId::of::<E>())
+            .unwrap_or_else(|| {
+                panic!(
+                    "Cannot observe event before it is registered with init_component: {}",
+                    std::any::type_name::<E>(),
+                )
+            });
+        descriptor.events.push(event);
+
+        Self {
+            commands,
+            descriptor,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Constructs an [`ObserverBuilder`] with a dynamic event id.
+    /// # Safety
+    /// Caller must ensure that the component associated with `id` is accessible as E
+    #[must_use]
+    pub unsafe fn new_with_id(event: ComponentId, commands: Commands<'w, 'w>) -> Self {
+        let mut descriptor = ObserverDescriptor::default();
+        descriptor.events.push(event);
+
+        Self {
+            commands,
+            descriptor,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Adds `NewE` to the list of events listened to by this observer.
+    /// Observers that listen to multiple types of events can no longer access the typed event data.
+    pub fn on_event<NewE: 'static>(&mut self) -> &mut ObserverBuilder<'w, ()> {
+        let event = self
+            .commands
+            .components()
+            .get_id(TypeId::of::<NewE>())
+            .unwrap_or_else(|| {
+                panic!(
+                    "Cannot observe event before it is registered with init_component: {}",
+                    std::any::type_name::<NewE>(),
+                )
+            });
+        self.descriptor.events.push(event);
+        // SAFETY: () will not allow bad memory access as it has no size
+        unsafe { std::mem::transmute(self) }
+    }
+
+    /// Add `events` to the list of events listened to by this observer.
+    /// Observers that listen to multiple types of events can no longer access the typed event data.
```

Comment:
Wouldn't hurt to be more explicit about what happens when the `events` iterator yields exactly `1` item & the observer listens to no other events.